## HatespeechClassification_with_ML

ဒီ tutorial က myHatespeech ဒေတာကို သုံးပြီး traditional ML approach တွေနဲ့ hatespeech classification ကို AI Engineering (Fundamental) အတန်းက ကျောင်းသားတွေကို လက်တွေ့ လုပ်ပြခဲ့တာပါ။

လုပ်ပြရင်းနဲ့ လက်ရှိ ရှဲထားတဲ့ myHatespeech ဒေတာထဲမှာ တချို့ လေဘယ်မှားနေတာတွေကို ရှာဖွေ့ တွေ့ရှိခဲ့ရတယ်။ ပြင်ထားတဲ့ myHatespeech ဒေတာကိုလည်း myHatespeech GitHub မှာ update ပြန်လုပ်ပေးထားပါမယ်။

By Ye Kyaw Thu, Lab Leader, Language Understanding Lab, Myanmar.

Date: 4 Oct 2026  

==========

myHatespeech Github Link: https://github.com/ye-kyaw-thu/myHateSpeech  
Long format data: https://github.com/ye-kyaw-thu/myHateSpeech/tree/main/corpus/version1.0/experiment/xgboost/long-data

Reference Code: https://github.com/ye-kyaw-thu/kh-polarity

## Change Fasttext Format into CSV 

In [3]:
%cd /mnt/disk1/ye/demo/hs/data/

/mnt/disk1/ye/demo/hs/data


In [4]:
!wc *

   1000   17892  223453 ltest.txt
   9137  163482 2068583 ltrain.txt
  10137  181374 2292036 total


In [5]:
!head -n 3 *.txt

==> ltest.txt <==
__label__ab $ရူး ဒိုင် တွေ နဲ့ တော့ ဟူး 🥶🥶🥶🥵🥵🥵
__label__bo ဘာ မ ဟုတ် တာ လို့ ပြော ထွက် တဲ့ ပါးစပ် ပေါက် ရေမြင်းပါးစပ် ကျပ်မပြည့်မ 😒 ငါ ပြော မိ ပြန် ပီ 😔
__label__ab စောက်ကောင်မ မ သိ နဲ့

==> ltrain.txt <==
__label__ab လီး တောင် ရိုးရိုး မ ဟုတ် ဘူး တပတ်လည် လီး ပဲ
__label__ab ဆူ စရာ ရှိ လည်း ကွယ်ရာ မှာ ပြော ပါ လား ကွာ စိတ် နု တဲ့ သူ ဆို ရှက် ပြီး သတ်သေ လောက် တယ် တစ် နိုင်ငံ လုံး မြင် အောင် ပြ နေ ကြ တာ မှား ရင် လည်း မှား မယ် အမှား ဆို အမှန် ကို ထောက်ပြ ပေး မှ ဆရာ ဖြစ် မှာ ခု ဟာ က ဆောက်ရှက်ခွဲ နေ တာ
__label__ab သေခါနီး ရိက္ခာ ယူ နေ ကြ တယ် GL


## Prepare Format Conversion Code with DeepSeek

Example Prompt: 

```
I have following data format:

ye@lst-hpc3090:/mnt/disk1/ye/demo/hs/data$ head ltrain.txt
__label__ab လီး တောင် ရိုးရိုး မ ဟုတ် ဘူး တပတ်လည် လီး ပဲ
__label__ab ဆူ စရာ ရှိ လည်း ကွယ်ရာ မှာ ပြော ပါ လား ကွာ စိတ် နု တဲ့ သူ ဆို ရှက် ပြီး သတ်သေ လောက် တယ် တစ် နိုင်ငံ လုံး မြင် အောင် ပြ နေ ကြ တာ မှား ရင် လည်း မှား မယ် အမှား ဆို အမှန် ကို ထောက်ပြ ပေး မှ ဆရာ ဖြစ် မှာ ခု ဟာ က ဆောက်ရှက်ခွဲ နေ တာ
__label__ab သေခါနီး ရိက္ခာ ယူ နေ ကြ တယ် GL
__label__ab အောက်တန်းစား စိတ်ဓာတ် တွေ ဖာဆန် တဲ့ လုပ်ရပ် နဲ့ အတွေး တွေ
__label__no စာအုပ် အဟောင်း သတင်းစာ အဟောင်း ပဲ ကျန် တော့ မှာ ဗျာ ရောင်းစား စရာ က လည်း
__label__ab ဖာ ဖြစ် မယ့် ကျောင်းသူ လေး များ တွေ့ ရှိ
__label__ab ဒီ khwayဝဲစား ကောင် ကို ထောင် ချ ပစ် ရ မှာ
__label__ab ရူး ရင် လည်း ထမင်းလွတ်ရူး ကြ ပါ ကွာ
__label__bo နင့် အသံ နဲ့ သီချင်း နဲ့ နင့် ဟန်ပန် အမူရာ က တခြားစီ ကားယား ပွတ်သပ် နေ တာ မသာမ ရဲ့ မြူဆွယ် နေ တာ လား
__label__ab လီး ပဲ ဟေ့
ye@lst-hpc3090:/mnt/disk1/ye/demo/hs/data$ head ltest.txt
__label__ab $ရူး ဒိုင် တွေ နဲ့ တော့ ဟူး 🥶🥶🥶🥵🥵🥵
__label__bo ဘာ မ ဟုတ် တာ လို့ ပြော ထွက် တဲ့ ပါးစပ် ပေါက် ရေမြင်းပါးစပ် ကျပ်မပြည့်မ 😒 ငါ ပြော မိ ပြန် ပီ 😔
__label__ab စောက်ကောင်မ မ သိ နဲ့
__label__ab mmsp ပဲ ကိုမေကိုလိုး တွေ တစ်နေကုန် ပျက် လာ တော့ လည်း လာ လိုက် ပျက် လိုက် နဲ့
__label__ab ဘောပြားမကြီး ဟ
__label__ab ဘောမ ခွေး ကြီး တက် လာ တယ် သတိထား ဟေ့
__label__ab ဒီ ပြိုင်ပွဲ ကြည့် ရ တာ ပြိုင်ပွဲ ဝင် တဲ့ ညီ တွေ ညီမ တွေ ကို သနား တယ် တစ် နိုင်ငံ လုံး ရှေ့ မှာ အဆူ အဆဲ ခံ စောက်ရှက်ခွဲ နေ တဲ့ အတိုင်း ပဲ Model ဖြစ် မ ဖြစ် တော့ မ သိ ဘူး 😒😥😡😡😡👿
__label__ab မအေလိုး လို့ မြင် ပါ တယ် ။
__label__ab စောက်ခွက် က လူ နဲ့ မ တူ
__label__ab အဲ့ ဆို ရင် ဖင်ခံ လိုက် ပါ ညီ 🔥🍆
ye@lst-hpc3090:/mnt/disk1/ye/demo/hs/data$

==========

I wanna change above file format into CSV format as follows:
text,label
$ရူး ဒိုင် တွေ နဲ့ တော့ ဟူး 🥶🥶🥶🥵🥵🥵,ab
ဘာ မ ဟုတ် တာ လို့ ပြော ထွက် တဲ့ ပါးစပ် ပေါက် ရေမြင်းပါးစပ် ကျပ်မပြည့်မ 😒 ငါ ပြော မိ ပြန် ပီ 😔,bo
စောက်ကောင်မ မ သိ နဲ့,ab
ပဲ ကိုမေကိုလိုး တွေ တစ်နေကုန် ပျက် လာ တော့ လည်း လာ လိုက် ပျက် လိုက် နဲ့,ab
ဘောပြားမကြီး ဟ,ab
ဘောမ ခွေး ကြီး တက် လာ တယ် သတိထား ဟေ့,ab
ဒီ ပြိုင်ပွဲ ကြည့် ရ တာ ပြိုင်ပွဲ ဝင် တဲ့ ညီ တွေ ညီမ တွေ ကို သနား တယ် တစ် နိုင်ငံ လုံး ရှေ့ မှာ အဆူ အဆဲ ခံ စောက်ရှက်ခွဲ နေ တဲ့ အတိုင်း ပဲ Model ဖြစ် မ ဖြစ် တော့ မ သိ ဘူး 😒😥😡😡😡👿,ab
မအေလိုး လို့ မြင် ပါ တယ် ။,ab
စောက်ခွက် က လူ နဲ့ မ တူ,ab
အဲ့ ဆို ရင် ဖင်ခံ လိုက် ပါ ညီ 🔥🍆,ab

============

Help me to write python code for conversion from fasttext to CSV? I wanna you to add some necessary command line arguments such as --csv_filename or -c, --fasttext_filename or -f, --help.
I need complete python code.
```

In [7]:
%cd /mnt/disk1/ye/demo/hs/

/mnt/disk1/ye/demo/hs


In [8]:
!cat ./fasttext_to_csv.py

#!/usr/bin/env python3
"""
Convert a fastText-format file (e.g. `__label__ab some text here`)
into a CSV file with two columns: text,label

Usage examples:
    python fasttext_to_csv.py -f ltrain.txt -c ltrain.csv
    python fasttext_to_csv.py --fasttext_filename ltest.txt --csv_filename ltest.csv
"""

import argparse
import csv
import sys

LABEL_PREFIX = "__label__"


def parse_fasttext_line(line: str):
    """
    Parse a single line of fastText data.

    Returns a tuple (labels, text) where:
        labels : list of label strings (without the __label__ prefix)
        text   : the remaining text
    Returns None if the line is empty or contains no label.
    """
    line = line.rstrip("\r\n")
    if not line.strip():
        return None

    labels = []
    # Extract all leading __label__ tokens
    while line.startswith(LABEL_PREFIX):
        parts = line.split(" ", 1)
        label_token = parts[0]
        labels.append(label_token[len(LABEL_PREFIX):])
        line = parts[1] if 

## Format Conversion

In [9]:
!python ./fasttext_to_csv.py -f ./data/ltrain.txt -c ./data/ltrain.csv

[OK] Read 9137 lines, wrote 9137 rows to './data/ltrain.csv'.


In [10]:
!python ./fasttext_to_csv.py -f ./data/ltest.txt -c ./data/ltest.csv

[OK] Read 1000 lines, wrote 1000 rows to './data/ltest.csv'.


In [11]:
!head ./data/ltrain.csv

text,label
လီး တောင် ရိုးရိုး မ ဟုတ် ဘူး တပတ်လည် လီး ပဲ,ab
ဆူ စရာ ရှိ လည်း ကွယ်ရာ မှာ ပြော ပါ လား ကွာ စိတ် နု တဲ့ သူ ဆို ရှက် ပြီး သတ်သေ လောက် တယ် တစ် နိုင်ငံ လုံး မြင် အောင် ပြ နေ ကြ တာ မှား ရင် လည်း မှား မယ် အမှား ဆို အမှန် ကို ထောက်ပြ ပေး မှ ဆရာ ဖြစ် မှာ ခု ဟာ က ဆောက်ရှက်ခွဲ နေ တာ,ab
သေခါနီး ရိက္ခာ ယူ နေ ကြ တယ် GL,ab
အောက်တန်းစား စိတ်ဓာတ် တွေ ဖာဆန် တဲ့ လုပ်ရပ် နဲ့ အတွေး တွေ,ab
စာအုပ် အဟောင်း သတင်းစာ အဟောင်း ပဲ ကျန် တော့ မှာ ဗျာ ရောင်းစား စရာ က လည်း,no
ဖာ ဖြစ် မယ့် ကျောင်းသူ လေး များ တွေ့ ရှိ,ab
ဒီ khwayဝဲစား ကောင် ကို ထောင် ချ ပစ် ရ မှာ,ab
ရူး ရင် လည်း ထမင်းလွတ်ရူး ကြ ပါ ကွာ,ab
နင့် အသံ နဲ့ သီချင်း နဲ့ နင့် ဟန်ပန် အမူရာ က တခြားစီ ကားယား ပွတ်သပ် နေ တာ မသာမ ရဲ့ မြူဆွယ် နေ တာ လား,bo


In [12]:
!head ./data/ltest.csv

text,label
$ရူး ဒိုင် တွေ နဲ့ တော့ ဟူး 🥶🥶🥶🥵🥵🥵,ab
ဘာ မ ဟုတ် တာ လို့ ပြော ထွက် တဲ့ ပါးစပ် ပေါက် ရေမြင်းပါးစပ် ကျပ်မပြည့်မ 😒 ငါ ပြော မိ ပြန် ပီ 😔,bo
စောက်ကောင်မ မ သိ နဲ့,ab
mmsp ပဲ ကိုမေကိုလိုး တွေ တစ်နေကုန် ပျက် လာ တော့ လည်း လာ လိုက် ပျက် လိုက် နဲ့,ab
ဘောပြားမကြီး ဟ,ab
ဘောမ ခွေး ကြီး တက် လာ တယ် သတိထား ဟေ့,ab
ဒီ ပြိုင်ပွဲ ကြည့် ရ တာ ပြိုင်ပွဲ ဝင် တဲ့ ညီ တွေ ညီမ တွေ ကို သနား တယ် တစ် နိုင်ငံ လုံး ရှေ့ မှာ အဆူ အဆဲ ခံ စောက်ရှက်ခွဲ နေ တဲ့ အတိုင်း ပဲ Model ဖြစ် မ ဖြစ် တော့ မ သိ ဘူး 😒😥😡😡😡👿,ab
မအေလိုး လို့ မြင် ပါ တယ် ။,ab
စောက်ခွက် က လူ နဲ့ မ တူ,ab


## Downloading Codes

In [13]:
!wget https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/dtree-classifier.py

--2026-10-04 15:40:08--  https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/dtree-classifier.py
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/dtree-classifier.py [following]
--2026-10-04 15:40:08--  https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/dtree-classifier.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6615 (6.5K) [text/plain]
Saving to: ‘dtree-classifier.py’

dtree-classifier.py 100%[===================>]   6.46K  --.-KB/s    in 0s      

2026-10-04 15:40:09 (58.9 MB/s) - ‘dtree-classif

In [14]:
!wget https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/knn-classifier.py

--2026-10-04 15:41:34--  https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/knn-classifier.py
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/knn-classifier.py [following]
--2026-10-04 15:41:34--  https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/knn-classifier.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6639 (6.5K) [text/plain]
Saving to: ‘knn-classifier.py’

knn-classifier.py   100%[===================>]   6.48K  --.-KB/s    in 0s      

2026-10-04 15:41:34 (58.4 MB/s) - ‘knn-classifier.py’ sa

In [15]:
!wget https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/rforest-classifier.py

--2026-10-04 15:42:03--  https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/rforest-classifier.py
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/rforest-classifier.py [following]
--2026-10-04 15:42:03--  https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/rforest-classifier.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6675 (6.5K) [text/plain]
Saving to: ‘rforest-classifier.py’

rforest-classifier. 100%[===================>]   6.52K  --.-KB/s    in 0s      

2026-10-04 15:42:04 (52.1 MB/s) - ‘rfore

In [16]:
!wget https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/sgd-classifier2.py

--2026-10-04 15:42:49--  https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/sgd-classifier2.py
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/sgd-classifier2.py [following]
--2026-10-04 15:42:49--  https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/sgd-classifier2.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6519 (6.4K) [text/plain]
Saving to: ‘sgd-classifier2.py’

sgd-classifier2.py  100%[===================>]   6.37K  --.-KB/s    in 0s      

2026-10-04 15:42:50 (58.9 MB/s) - ‘sgd-classifier2.p

In [17]:
!wget https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/svm-classifier.py

--2026-10-04 15:43:20--  https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/svm-classifier.py
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/svm-classifier.py [following]
--2026-10-04 15:43:21--  https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/svm-classifier.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6489 (6.3K) [text/plain]
Saving to: ‘svm-classifier.py’

svm-classifier.py   100%[===================>]   6.34K  --.-KB/s    in 0s      

2026-10-04 15:43:22 (58.1 MB/s) - ‘svm-classifier.py’ sa

In [18]:
!wget https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/sgd-tune-classifier.py

--2026-10-04 15:43:54--  https://github.com/ye-kyaw-thu/kh-polarity/raw/refs/heads/main/code/sgd-tune-classifier.py
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/sgd-tune-classifier.py [following]
--2026-10-04 15:43:54--  https://raw.githubusercontent.com/ye-kyaw-thu/kh-polarity/refs/heads/main/code/sgd-tune-classifier.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7949 (7.8K) [text/plain]
Saving to: ‘sgd-tune-classifier.py’

sgd-tune-classifier 100%[===================>]   7.76K  --.-KB/s    in 0s      

2026-10-04 15:43:55 (32.3 MB/s) - ‘s

## Check/Learn Traditional ML Codes

In [19]:
!cat ./knn-classifier.py

## KNN Classifier for Khmer Polarity
## Written by Ye Kyaw Thu, 
## Affiliate Professor, IDRI, CADT, Cambodia
## Used for 4th NLP/AI Workshop, Chiang Mai, Experiment
## Last updated: 8 Feb 2023
## Reference:
## https://towardsdatascience.com/building-a-sentiment-classifier-using-scikit-learn-54c8e7c5d2f0
## https://vitalflux.com/accuracy-precision-recall-f1-score-python-example/  
## https://stackoverflow.com/questions/62792001/precision-and-recall-are-the-same-within-a-model
## https://towardsdatascience.com/micro-macro-weighted-averages-of-f1-score-clearly-explained-b603420b292f
## https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

import pandas as pd
import re
from os import system, listdir
from os.path import isfile, join
from random import shuffle

import warnings
warnings.filterwarnings("ignore")


polar_train = pd.read_csv('csv/train.csv')
polar_test = pd.read_csv('csv/test.csv')

### Text Vectorization

from sklearn.feature_extraction.

In [20]:
!cat dtree-classifier.py

## Decision Tree Classifier for Khmer Polarity
## Written by Ye Kyaw Thu, 
## Affiliate Professor, IDRI, CADT, Cambodia
## Used for 4th NLP/AI Workshop, Chiang Mai, Experiment
## Last updated: 8 Feb 2023
## Reference:
## https://towardsdatascience.com/building-a-sentiment-classifier-using-scikit-learn-54c8e7c5d2f0
## https://vitalflux.com/accuracy-precision-recall-f1-score-python-example/  
## https://stackoverflow.com/questions/62792001/precision-and-recall-are-the-same-within-a-model
## https://towardsdatascience.com/micro-macro-weighted-averages-of-f1-score-clearly-explained-b603420b292f
## https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

import pandas as pd
import re
from os import system, listdir
from os.path import isfile, join
from random import shuffle

import warnings
warnings.filterwarnings("ignore")


polar_train = pd.read_csv('csv/train.csv')
polar_test = pd.read_csv('csv/test.csv')

### Text Vectorization

from sklearn.feature_e

In [21]:
!cat sgd-classifier2.py

## Stochastic Gradient Descent Classifier for Khmer Polarity
## Written by Ye Kyaw Thu, 
## Affiliate Professor, IDRI, CADT, Cambodia
## Used for 4th NLP/AI Workshop, Chiang Mai, Experiment
## Last updated: 8 Feb 2023
## Reference:
## https://towardsdatascience.com/building-a-sentiment-classifier-using-scikit-learn-54c8e7c5d2f0
## https://vitalflux.com/accuracy-precision-recall-f1-score-python-example/  
## https://stackoverflow.com/questions/62792001/precision-and-recall-are-the-same-within-a-model
## https://towardsdatascience.com/micro-macro-weighted-averages-of-f1-score-clearly-explained-b603420b292f
## https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

import pandas as pd
import re
from os import system, listdir
from os.path import isfile, join
from random import shuffle

import warnings
warnings.filterwarnings("ignore")


polar_train = pd.read_csv('csv/train.csv')
polar_test = pd.read_csv('csv/test.csv')

### Text Vectorization

from skl

In [22]:
!cat sgd-tune-classifier.py

## Original code from:
## https://towardsdatascience.com/building-a-sentiment-classifier-using-scikit-learn-54c8e7c5d2f0
## Used for Khmer Polarity Classification Baseline
## Small modification by Ye Kyaw Thu, IDRI, CADT, Cambodia
## Last updated: 8 Feb 2023
## Reference:
## https://towardsdatascience.com/building-a-sentiment-classifier-using-scikit-learn-54c8e7c5d2f0
## https://vitalflux.com/accuracy-precision-recall-f1-score-python-example/  
## https://stackoverflow.com/questions/62792001/precision-and-recall-are-the-same-within-a-model
## https://towardsdatascience.com/micro-macro-weighted-averages-of-f1-score-clearly-explained-b603420b292f
## https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

import pandas as pd
import re
from os import system, listdir
from os.path import isfile, join
from random import shuffle

import warnings
warnings.filterwarnings("ignore")

#system('wget "http://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz

In [23]:
!cat svm-classifier.py

## Support Vector Machine Classifier for Khmer Polarity
## Written by Ye Kyaw Thu, 
## Affiliate Professor, IDRI, CADT, Cambodia
## Used for 4th NLP/AI Workshop, Chiang Mai, Experiment
## Last updated: 8 Feb 2023
## Reference:
## https://towardsdatascience.com/building-a-sentiment-classifier-using-scikit-learn-54c8e7c5d2f0
## https://vitalflux.com/accuracy-precision-recall-f1-score-python-example/  
## https://stackoverflow.com/questions/62792001/precision-and-recall-are-the-same-within-a-model
## https://towardsdatascience.com/micro-macro-weighted-averages-of-f1-score-clearly-explained-b603420b292f
## https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

import pandas as pd
import re
from os import system, listdir
from os.path import isfile, join
from random import shuffle

import warnings
warnings.filterwarnings("ignore")


polar_train = pd.read_csv('csv/train.csv')
polar_test = pd.read_csv('csv/test.csv')

### Text Vectorization

from sklearn.

In [24]:
!cat rforest-classifier.py

## Random Forest Classifier for Khmer Polarity
## Written by Ye Kyaw Thu, 
## Affiliate Professor, IDRI, CADT, Cambodia
## Used for 4th NLP/AI Workshop, Chiang Mai, Experiment
## Last updated: 8 Feb 2023
## Reference:
## https://towardsdatascience.com/building-a-sentiment-classifier-using-scikit-learn-54c8e7c5d2f0
## https://vitalflux.com/accuracy-precision-recall-f1-score-python-example/  
## https://stackoverflow.com/questions/62792001/precision-and-recall-are-the-same-within-a-model
## https://towardsdatascience.com/micro-macro-weighted-averages-of-f1-score-clearly-explained-b603420b292f
## https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

import pandas as pd
import re
from os import system, listdir
from os.path import isfile, join
from random import shuffle

import warnings
warnings.filterwarnings("ignore")


polar_train = pd.read_csv('csv/train.csv')
polar_test = pd.read_csv('csv/test.csv')

### Text Vectorization

from sklearn.feature_e

## Hatespeech Classification Experiment with Traditional ML Approaches

In [25]:
%pwd

'/mnt/disk1/ye/demo/hs'

In [26]:
!mkdir csv

In [27]:
!cp ./data/*.csv ./csv/

In [28]:
!wc ./csv/*.csv

   1001   16893  215482 ./csv/ltest.csv
   9138  154346 1995558 ./csv/ltrain.csv
  10139  171239 2211040 total


In [29]:
!mv ./csv/ltrain.csv ./csv/train.csv

In [30]:
!mv ./csv/ltest.csv ./csv/test.csv

## KNN

In [31]:
!time python ./knn-classifier.py 

Traceback (most recent call last):
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 126, in <module>
    train_and_show_scores_KNN(X_train_unigram, y_train, 'KNN, Unigram Counts', 'knn_unigram_count.joblib')
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 111, in train_and_show_scores_KNN
    X_train, X_valid, y_train, y_valid = train_test_split(
                                         ^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/utils/_param_validation.py", line 218, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2940, in train_test_split
    train, test = next(cv.split(X=arrays[0], y=stratify))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2429, in split
    y = check_array(y, input_name="y", ensure_2d=False

## Errors in Original Fasttxt Format Data

အော်ရဂျင်နယ် ရှဲထားတဲ့ myHatespeech Data တွေမှာ အထူးသဖြင့် လေဘယ် အမှားတွေရှိတာ၊ ပြီးတော့ မြန်မာစာ စာကြောင်းတွေထဲမှာ ကော်မာတွေ ပါနေတာကို တွေ့ရှိခဲ့ရတယ်။ အဲဒါကြောင့် ML code တွေကို run တဲ့အခါမှာ error တက်တာပါ။

In [32]:
%pwd

'/mnt/disk1/ye/demo/hs'

## Format Conversion AGAIN

In [33]:
!python ./fasttext_to_csv.py --help

usage: fasttext_to_csv.py [-h] -f FASTTEXT_FILENAME -c CSV_FILENAME
                          [--label_sep LABEL_SEP] [--no_header]

Convert a fastText-format file into a CSV file with columns 'text' and
'label'.

options:
  -h, --help            show this help message and exit
  -f FASTTEXT_FILENAME, --fasttext_filename FASTTEXT_FILENAME
                        Path to the input fastText file (e.g. ltrain.txt).
  -c CSV_FILENAME, --csv_filename CSV_FILENAME
                        Path to the output CSV file (e.g. ltrain.csv).
  --label_sep LABEL_SEP
                        Separator used when a line has multiple labels
                        (default: space).
  --no_header           Do not write the 'text,label' header row.


In [34]:
!python ./fasttext_to_csv.py -f ./clean_data/train.fasttext -c ./csv/train.csv

[OK] Read 9137 lines, wrote 9137 rows to './csv/train.csv'.


In [35]:
!python ./fasttext_to_csv.py -f ./clean_data/test.fasttext -c ./csv/test.csv

[OK] Read 1000 lines, wrote 1000 rows to './csv/test.csv'.


## Classification with Traditional ML Approaches

### KNN

In [36]:
!time python ./knn-classifier.py 

mkdir: cannot create directory ‘data_preprocessors’: File exists
mkdir: cannot create directory ‘vectorized_data’: File exists
Traceback (most recent call last):
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 126, in <module>
    train_and_show_scores_KNN(X_train_unigram, y_train, 'KNN, Unigram Counts', 'knn_unigram_count.joblib')
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 111, in train_and_show_scores_KNN
    X_train, X_valid, y_train, y_valid = train_test_split(
                                         ^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/utils/_param_validation.py", line 218, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2940, in train_test_split
    train, test = next(cv.split(X=arrays[0], y=stratify))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.

## Check CSV Files

In [37]:
!python ./check_csv.py -i ./csv/train.csv


File          : ./csv/train.csv
Total lines   : 9138
Blank lines   : 0
>1 comma      : 0
>2 fields     : 0


In [38]:
!python ./check_csv.py -i ./csv/test.csv


File          : ./csv/test.csv
Total lines   : 1001
Blank lines   : 0
>1 comma      : 0
>2 fields     : 0


In [39]:
!time python ./knn-classifier.py 

Traceback (most recent call last):
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 126, in <module>
    train_and_show_scores_KNN(X_train_unigram, y_train, 'KNN, Unigram Counts', 'knn_unigram_count.joblib')
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 111, in train_and_show_scores_KNN
    X_train, X_valid, y_train, y_valid = train_test_split(
                                         ^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/utils/_param_validation.py", line 218, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2940, in train_test_split
    train, test = next(cv.split(X=arrays[0], y=stratify))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2429, in split
    y = check_array(y, input_name="y", ensure_2d=False

## Debugging or Checking Data Error

In [40]:
!cat debug_nan.py

#!/usr/bin/env python3
import pandas as pd

for path in ["csv/train.csv", "csv/test.csv"]:
    print("=" * 70)
    print("FILE:", path)
    df = pd.read_csv(path)                 # <-- default: na_values applied
    print("shape         :", df.shape)
    print("dtypes        :")
    print(df.dtypes)
    print("NaN in text   :", df["text"].isna().sum())
    print("NaN in label  :", df["label"].isna().sum())
    print("Unique labels :", df["label"].unique()[:30])

    bad = df[df["label"].isna()]
    if not bad.empty:
        print("\n>> Rows with NaN label:")
        print(bad.head(20).to_string())


In [41]:
!python ./debug_nan.py

FILE: csv/train.csv
shape         : (9137, 2)
dtypes        :
text     str
label    str
dtype: object
NaN in text   : 0
NaN in label  : 3
Unique labels : <ArrowStringArray>
['ab', 'no', 'bo', 're', 'po', 'le', 'ed', 'se', 'ra', nan]
Length: 10, dtype: str

>> Rows with NaN label:
                                                                                     text label
1562                                                       ခွေးသား များ အက ကောင်း သွား တာ   NaN
2415                                      ယီး စား လွန်း လို့ ချီးမွမ်း နေ တာ ကို ဟုတ် တယ်   NaN
5049  စောက်ရူး ပဲ မြန်မာ နိုင်ငံ သား တစ် ယောက် အနေနဲ့ ငါ က တော့ လီး ပဲ မင်း လုပ်ရပ် တွေ က   NaN
FILE: csv/test.csv
shape         : (1000, 2)
dtypes        :
text     str
label    str
dtype: object
NaN in text   : 0
NaN in label  : 0
Unique labels : <ArrowStringArray>
['ab', 'bo', 'po', 'no', 'ed', 'le', 'se', 'ra', 're']
Length: 9, dtype: str


## Fixed Data Error

```
1563
__label__ ခွေးသား များ အက ကောင်း သွား တာ

2416
__label__ ယီး စား လွန်း လို့ ချီးမွမ်း နေ တာ ကို ဟုတ် တယ်

5050
__label__ စောက်ရူး ပဲ မြန်မာ နိုင်ငံ သား တစ် ယောက် အနေနဲ့ ငါ က တော့ လီး ပဲ မင်း လုပ်ရပ် တွေ က

__label__ed စောက်ရူး ပဲ မြန်မာ နိုင်ငံ သား တစ် ယောက် အနေနဲ့ ငါ က တော့ လီး ပဲ မင်း လုပ်ရပ် တွေ က
```

## Convert FastText Format into CSV AGAIN

In [42]:
%pwd

'/mnt/disk1/ye/demo/hs'

In [43]:
!python ./fasttext_to_csv.py -f ./clean_data/train.fasttext -c ./csv/train.csv

[OK] Read 9137 lines, wrote 9137 rows to './csv/train.csv'.


In [45]:
!wc ./csv/*.csv

   1001   16890  215467 ./csv/test.csv
   9138  154339 1995526 ./csv/train.csv
  10139  171229 2210993 total


In [46]:
!head ./csv/*.csv

==> ./csv/test.csv <==
text,label
$ရူး ဒိုင် တွေ နဲ့ တော့ ဟူး 🥶🥶🥶🥵🥵🥵,ab
ဘာ မ ဟုတ် တာ လို့ ပြော ထွက် တဲ့ ပါးစပ် ပေါက် ရေမြင်းပါးစပ် ကျပ်မပြည့်မ 😒 ငါ ပြော မိ ပြန် ပီ 😔,bo
စောက်ကောင်မ မ သိ နဲ့,ab
mmsp ပဲ ကိုမေကိုလိုး တွေ တစ်နေကုန် ပျက် လာ တော့ လည်း လာ လိုက် ပျက် လိုက် နဲ့,ab
ဘောပြားမကြီး ဟ,ab
ဘောမ ခွေး ကြီး တက် လာ တယ် သတိထား ဟေ့,ab
ဒီ ပြိုင်ပွဲ ကြည့် ရ တာ ပြိုင်ပွဲ ဝင် တဲ့ ညီ တွေ ညီမ တွေ ကို သနား တယ် တစ် နိုင်ငံ လုံး ရှေ့ မှာ အဆူ အဆဲ ခံ စောက်ရှက်ခွဲ နေ တဲ့ အတိုင်း ပဲ Model ဖြစ် မ ဖြစ် တော့ မ သိ ဘူး 😒😥😡😡😡👿,ab
မအေလိုး လို့ မြင် ပါ တယ် ။,ab
စောက်ခွက် က လူ နဲ့ မ တူ,ab

==> ./csv/train.csv <==
text,label
လီး တောင် ရိုးရိုး မ ဟုတ် ဘူး တပတ်လည် လီး ပဲ,ab
ဆူ စရာ ရှိ လည်း ကွယ်ရာ မှာ ပြော ပါ လား ကွာ စိတ် နု တဲ့ သူ ဆို ရှက် ပြီး သတ်သေ လောက် တယ် တစ် နိုင်ငံ လုံး မြင် အောင် ပြ နေ ကြ တာ မှား ရင် လည်း မှား မယ် အမှား ဆို အမှန် ကို ထောက်ပြ ပေး မှ ဆရာ ဖြစ် မှာ ခု ဟာ က ဆောက်ရှက်ခွဲ နေ တာ,ab
သေခါနီး ရိက္ခာ ယူ နေ ကြ တယ် GL,ab
အောက်တန်းစား စိတ်ဓာတ် တွေ ဖာဆန် တဲ့ လုပ်ရပ် နဲ့ အတွေး တွေ,ab
စာအုပ် အဟောင်း သတင်းစာ အ

## Classification with KNN

In [47]:
!time python ./knn-classifier.py

mkdir: cannot create directory ‘data_preprocessors’: File exists
mkdir: cannot create directory ‘vectorized_data’: File exists
Traceback (most recent call last):
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 126, in <module>
    train_and_show_scores_KNN(X_train_unigram, y_train, 'KNN, Unigram Counts', 'knn_unigram_count.joblib')
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 111, in train_and_show_scores_KNN
    X_train, X_valid, y_train, y_valid = train_test_split(
                                         ^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/utils/_param_validation.py", line 218, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2940, in train_test_split
    train, test = next(cv.split(X=arrays[0], y=stratify))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.

## Debugging

In [48]:
!python ./debug_nan.py

FILE: csv/train.csv
shape         : (9137, 2)
dtypes        :
text     str
label    str
dtype: object
NaN in text   : 0
NaN in label  : 1
Unique labels : <ArrowStringArray>
['ab', 'no', 'bo', 're', 'po', 'le', 'ed', 'se', 'ra', nan]
Length: 10, dtype: str

>> Rows with NaN label:
                                                 text label
2415  ယီး စား လွန်း လို့ ချီးမွမ်း နေ တာ ကို ဟုတ် တယ်   NaN
FILE: csv/test.csv
shape         : (1000, 2)
dtypes        :
text     str
label    str
dtype: object
NaN in text   : 0
NaN in label  : 0
Unique labels : <ArrowStringArray>
['ab', 'bo', 'po', 'no', 'ed', 'le', 'se', 'ra', 're']
Length: 9, dtype: str


## Format Conversion AGAIN

In [53]:
!python ./fasttext_to_csv.py -f ./clean_data/train.fasttext -c ./csv/train.csv

[OK] Read 9137 lines, wrote 9137 rows to './csv/train.csv'.


## Classification with KNN

In [52]:
!time python ./knn-classifier.py

Traceback (most recent call last):
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 126, in <module>
    train_and_show_scores_KNN(X_train_unigram, y_train, 'KNN, Unigram Counts', 'knn_unigram_count.joblib')
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 111, in train_and_show_scores_KNN
    X_train, X_valid, y_train, y_valid = train_test_split(
                                         ^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/utils/_param_validation.py", line 218, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2944, in train_test_split
    return list(
           ^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2946, in <genexpr>
    (_safe_indexing(a, train), _safe_indexing(a, test)) for a in arrays
     ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/p

## Notes on Some Errors That We Found

### Test data Errors

```
   612   ဗမာအေးရပ်ကွက် ကို နေ့စဉ် မီး ဖျက် ထား တာ အခု မနက် ၉ နာရီ ကတည်းက မီး ပျက် ထား တာ အခု ထိ မ လာ သေး ဘူး နင် တို့ YESC ဟာ တွေ တော်တော် $တရားလွန် သွား ပြီ နော် နေ ပူ တော့ လည်း နေ ပူ လို့ မီး ဖြတ် ထား အခု မိုးရွာ တော့ မီး ပါ မ မြင် ရ တော့ ဘူး မီတာခ ကျတော့ အချိန်မှန် ကောက် မီး ကျ တော့ ၄ နာရီ ပျက် ၄ နာရီ လာ ဖို့ တောင် သေချာ မ ပေး နိုင် ဘူး ဆို တော့ YESC က ဟာ တွေ thay နေ ကြ တာ လား 😡 😡 😡 ပြည်သူ တွေ ကို မီး ပေး ရ မယ့် အလှည့် ကို မ ပေး ဘဲ နင် တို့ မီး ခိုး ရောင်း နေ တာ"
   613  ab
   
    665  no
   666
   667  po
   
__label__ animal abuse ပဲ အဲ့ လို ကောင် တွေ တရားစွဲ လို့ မ ရ ရင် ဝိုင်းy-te မှ
```

### Training Data Errors

```
4562
__label__ab ခွေးသား တွေ သင်္ကြန် အရင် က မှ ပုံမှန် ပျက် ပြီး ည ဘက် တွေ မ ပျက် ဘူး သင်္ကြန် ကာလ မှာ 4,5 ရက် လောက် မ ဖြတ် ဘဲ သင်္ကြန် ပြီး မှာ mmsp လို ဖြတ် နေ တယ် စောက်ခွေးသား တွေ ည ဘက် တော့ မ ဖြတ် နဲ့

__label__ab ခွေးသား တွေ သင်္ကြန် အရင် က မှ ပုံမှန် ပျက် ပြီး ည ဘက် တွေ မ ပျက် ဘူး သင်္ကြန် ကာလ မှာ 4 ရက် 5 ရက် လောက် မ ဖြတ် ဘဲ သင်္ကြန် ပြီး မှာ mmsp လို ဖြတ် နေ တယ် စောက်ခွေးသား တွေ ည ဘက် တော့ မ ဖြတ် နဲ့

6289
__label__le အလိုလို လမ်းဘေး ခွေး ဆို တာ အစာ တောင် ကောင်းကောင်း မ စား ရ ဘူး , သူ့ ခမျာ ကြောက်လန့် ခြင်း ငတ်ပြတ် ခြင်း နဲ့ ဘဝ ကို ဖြတ်သန်း နေ တာ , , အဲ့ တာ ကို အဲ့ မျိူး ခွေး နဲ့ အလကား နေ သွား ကိုက် ခိုင်း တယ် တိရိစ္ဆာန်ကောင် ရေ နင့် ကို အဲ့ ခွေး ကိုက် ပြီး သေပါစေ

__label__le အလိုလို လမ်းဘေး ခွေး ဆို တာ အစာ တောင် ကောင်းကောင်း မ စား ရ ဘူး သူ့ ခမျာ ကြောက်လန့် ခြင်း ငတ်ပြတ် ခြင်း နဲ့ ဘဝ ကို ဖြတ်သန်း နေ တာ အဲ့ တာ ကို အဲ့ မျိူး ခွေး နဲ့ အလကား နေ သွား ကိုက် ခိုင်း တယ် တိရိစ္ဆာန်ကောင် ရေ နင့် ကို အဲ့ ခွေး ကိုက် ပြီး သေပါစေ

7266
__label__po အဓိပ္ပာယ် က တော့ မြန်မာ မှာ ခွေး နီပိန်း ပဒက် တွေ အကြမ်းဖက် သလို တို့ နိုင်ငံ အဖြစ် ခံ မှာ လား တို့ မ လုပ် နိုင် ရင် တောင် ယ္ခု ဝန်ကြီးချုပ် , ဒု ဝန်ကြီးချုပ် ကို ဝန်းရံ ကြ ရ မယ် ခွေး နီပိန်း ပဒက် နဲ့ တူ တဲ့ ပီတာ အဖွဲ့ ကို လက် မ ခံ ကြောင်း

8613
__label__po ဒီ မအေloe အငုံ စိတ် လောက်ကောင် ပါတီ တွေ ကို NUG,NUCC က တရား မ ဝင် ကြေငြာ ပေး ပါ အကြမ်းဖက် စစ်တပ် နဲ့ ပူးပေါင်း သူ တွေ အဖြစ် နိုင်ငံတော် ပုန်ကန် မှု မြောက် သူ တွေ အဖြစ် ကြေငြာ ပေး ပါ ။

__label__po ဒီ မအေloe အငုံ စိတ် လောက်ကောင် ပါတီ တွေ ကို NUG NUCC က တရား မ ဝင် ကြေငြာ ပေး ပါ အကြမ်းဖက် စစ်တပ် နဲ့ ပူးပေါင်း သူ တွေ အဖြစ် နိုင်ငံတော် ပုန်ကန် မှု မြောက် သူ တွေ အဖြစ် ကြေငြာ ပေး ပါ ။

8975
__label__ab ခွေးသတောင်းစား YESC က ဟာ တွေ မီး ကို ဒေါပုံ , ဗမာအေး ကို လီး လို ပေး တချို့ မြိုနယ် လာ ချိန် ပျက် ချိန် မှန် ည ပိုင်း ဆို အကုန် မီး လာ တယ် ဒေါပုံ ပဲ မီး ဖြတ် ထား တာ အခု လည်း ည ၉ နာရီ လာ ရ မှာ နာရီဝက် ပဲ မီး ပေး ပြီး မီး ပျက် သွား ပြန် ပြီ ခွေးသတောင်းစား တွေ မီး ခိုး ရောင်း နေ တာ မသာ တွေ 😡 😡 😡 နေ့စဉ် တိုင်း တစ် နေ့ လုံး မှ ၃ နာရီ လောက် ပဲ မီး လာ တယ် ခွေးသတောင်းစား YESC ဟာ တွေ ရေ $သုံးမကျ တာ လွန် လွန်း တယ် 😡 $ဆံ ကျ ငမ်းငမ်းတက် 😡

__label__ab ခွေးသတောင်းစား YESC က ဟာ တွေ မီး ကို ဒေါပုံ ဗမာအေး ကို လီး လို ပေး တချို့ မြိုနယ် လာ ချိန် ပျက် ချိန် မှန် ည ပိုင်း ဆို အကုန် မီး လာ တယ် ဒေါပုံ ပဲ မီး ဖြတ် ထား တာ အခု လည်း ည ၉ နာရီ လာ ရ မှာ နာရီဝက် ပဲ မီး ပေး ပြီး မီး ပျက် သွား ပြန် ပြီ ခွေးသတောင်းစား တွေ မီး ခိုး ရောင်း နေ တာ မသာ တွေ 😡 😡 😡 နေ့စဉ် တိုင်း တစ် နေ့ လုံး မှ ၃ နာရီ လောက် ပဲ မီး လာ တယ် ခွေးသတောင်းစား YESC ဟာ တွေ ရေ $သုံးမကျ တာ လွန် လွန်း တယ် 😡 $ဆံ ကျ ငမ်းငမ်းတက် 😡

4244
__label__po ရူးလို့မဝသေးဘူးလား , တော်တော် စိတ်ကူးယဉ် တဲ့ သူ ပဲ , တိုင်းပြည် နာ လှ ပြီ , မြန်မြန် ဆင်း ပါ တော့

1563
__label__ ခွေးသား များ အက ကောင်း သွား တာ

2416
__label__ ယီး စား လွန်း လို့ ချီးမွမ်း နေ တာ ကို ဟုတ် တယ်

5050
__label__ စောက်ရူး ပဲ မြန်မာ နိုင်ငံ သား တစ် ယောက် အနေနဲ့ ငါ က တော့ လီး ပဲ မင်း လုပ်ရပ် တွေ က

__label__ed စောက်ရူး ပဲ မြန်မာ နိုင်ငံ သား တစ် ယောက် အနေနဲ့ ငါ က တော့ လီး ပဲ မင်း လုပ်ရပ် တွေ က
```

အထက်ပါ error တွေကို text editor မှာဖွင့်ပြီး manual correction လုပ်ခဲ့ပါတယ်။


## KNN Code Updated

In [54]:
!python ./knn-classifier.py

Train size: (9137, 2) | Test size: (1000, 2)
Train label distribution:
 label
ab    5115
no    1648
po     942
bo     478
le     286
se     267
ed     206
ra     105
re      90
Name: count, dtype: int64

Traceback (most recent call last):
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 87, in <module>
    train_and_show_scores_KNN(X_train_unigram,
  File "/mnt/disk1/ye/demo/hs/./knn-classifier.py", line 73, in train_and_show_scores_KNN
    X_train, X_valid, y_train, y_valid = train_test_split(
                                         ^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/utils/_param_validation.py", line 218, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2944, in train_test_split
    return list(
           ^^^^^
  File "/home/ye/.local/lib/python3.12/site-packages/sklearn/model_selection/_split.py", line 2946, in

## Code Updated

Error relating to python library version of Arrow, Panda etc.  

In [55]:
!time python ./knn-classifier.py

Train size: (9137, 2) | Test size: (1000, 2)
Train label distribution:
 label
ab    5115
no    1648
po     942
bo     478
le     286
se     267
ed     206
ra     105
re      90
Name: count, dtype: int64

KNN, Unigram Counts
Train score: 0.62 ; Validation score: 0.53

KNN, Unigram Tf-Idf
Train score: 0.64 ; Validation score: 0.56

KNN, Bigram Counts
Train score: 0.62 ; Validation score: 0.52

KNN, Bigram Tf-Idf
Train score: 0.6 ; Validation score: 0.54

KNN Test Result, Unigram Counts: 0.536
Error Rate: 0.46
              precision    recall  f1-score   support

          ab       0.58      0.88      0.70       550
          bo       0.43      0.09      0.15        64
          ed       0.00      0.00      0.00        21
          le       0.00      0.00      0.00        37
          no       0.27      0.19      0.22       186
          po       0.62      0.11      0.19        90
          ra       0.00      0.00      0.00        17
          re       0.00      0.00      0.00        11


## Updated KNN Code

In [56]:
!cat knn-classifier.py

#!/usr/bin/env python3
## KNN Classifier — fixed version
## - keep_default_na=False so labels like "NA"/"NaN" survive
## - .to_numpy() instead of .values to avoid PyArrow indexing bug
## - separate X_test_* variables for count vs tf-idf models

import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from os import system
from scipy.sparse import csr_matrix, save_npz
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from joblib import dump, load

# ---------------------------------------------------------------- load data
polar_train = pd.read_csv('csv/train.csv', keep_default_na=False)
polar_test  = pd.read_csv('csv/test.csv',  keep_default_na=False)

# Sanity check — abort early if labels still contain NaN
if polar_train['label'].isna().any() or polar_test['label'].isna().

## Updated Code for Decision Tree

In [57]:
!cat ./dtree-classifier.py

#!/usr/bin/env python3
## Decision Tree Classifier — fixed version
## - keep_default_na=False so labels like "NA"/"NaN" survive
## - .to_numpy() instead of .values to avoid PyArrow indexing bug
## - separate X_test_* variables for count vs tf-idf models
## - random_state=42 for reproducible splits

import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from os import system
from scipy.sparse import csr_matrix, save_npz
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.metrics import classification_report
from joblib import dump, load

# ---------------------------------------------------------------- load data
polar_train = pd.read_csv('csv/train.csv', keep_default_na=False)
polar_test  = pd.read_csv('csv/test.csv',  keep_default_na=False)

# Sanity check — abort early if labels still contain NaN
if polar_train['label'].isna().any() o

## Classification with Decision Tree Approach

In [58]:
!time python ./dtree-classifier.py

Train size: (9137, 2) | Test size: (1000, 2)
Train label distribution:
 label
ab    5115
no    1648
po     942
bo     478
le     286
se     267
ed     206
ra     105
re      90
Name: count, dtype: int64

DTREE, Unigram Counts
Train score: 0.92 ; Validation score: 0.5

DTREE, Unigram Tf-Idf
Train score: 0.92 ; Validation score: 0.51

DTREE, Bigram Counts
Train score: 0.92 ; Validation score: 0.52

DTREE, Bigram Tf-Idf
Train score: 0.92 ; Validation score: 0.52

Decision Tree Test Result, Unigram Counts: 0.522
Error Rate: 0.48
              precision    recall  f1-score   support

          ab       0.62      0.77      0.69       550
          bo       0.36      0.22      0.27        64
          ed       0.11      0.05      0.07        21
          le       0.17      0.08      0.11        37
          no       0.30      0.26      0.28       186
          po       0.46      0.29      0.35        90
          ra       0.33      0.06      0.10        17
          re       0.29      0.18   

## Updated Random Forest Code

In [59]:
!cat rforest-classifier.py

#!/usr/bin/env python3
## Random Forest Classifier — fixed version
## - keep_default_na=False so labels like "NA"/"NaN" survive
## - .to_numpy() instead of .values to avoid PyArrow indexing bug
## - separate X_test_* variables for count vs tf-idf models
## - random_state=42 for reproducible splits

import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from os import system
from scipy.sparse import csr_matrix, save_npz
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from joblib import dump, load

# ---------------------------------------------------------------- load data
polar_train = pd.read_csv('csv/train.csv', keep_default_na=False)
polar_test  = pd.read_csv('csv/test.csv',  keep_default_na=False)

# Sanity check — abort early if labels still contain NaN
if polar_tr

## Classification with Random Forest Approach

In [60]:
!time python ./rforest-classifier.py

Train size: (9137, 2) | Test size: (1000, 2)
Train label distribution:
 label
ab    5115
no    1648
po     942
bo     478
le     286
se     267
ed     206
ra     105
re      90
Name: count, dtype: int64

RFOREST, Unigram Counts
Train score: 0.92 ; Validation score: 0.57

RFOREST, Unigram Tf-Idf
Train score: 0.92 ; Validation score: 0.58

RFOREST, Bigram Counts
Train score: 0.92 ; Validation score: 0.58

RFOREST, Bigram Tf-Idf
Train score: 0.92 ; Validation score: 0.59

Random Forest Test Result, Unigram Counts: 0.576
Error Rate: 0.42
              precision    recall  f1-score   support

          ab       0.62      0.85      0.72       550
          bo       0.59      0.16      0.25        64
          ed       0.00      0.00      0.00        21
          le       0.33      0.05      0.09        37
          no       0.37      0.28      0.32       186
          po       0.64      0.40      0.49        90
          ra       0.00      0.00      0.00        17
          re       0.25    

## Updated Code for SGD

In [61]:
!cat ./sgd-classifier2.py

#!/usr/bin/env python3
## Stochastic Gradient Descent Classifier — fixed version
## - keep_default_na=False so labels like "NA"/"NaN" survive
## - .to_numpy() instead of .values to avoid PyArrow indexing bug
## - separate X_test_* variables for count vs tf-idf models
## - random_state=42 for reproducible splits

import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from os import system
from scipy.sparse import csr_matrix, save_npz
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import classification_report
from joblib import dump, load

# ---------------------------------------------------------------- load data
polar_train = pd.read_csv('csv/train.csv', keep_default_na=False)
polar_test  = pd.read_csv('csv/test.csv',  keep_default_na=False)

# Sanity check — abort early if labels still contain NaN
if

## Classification with SGD Approach

In [62]:
!time python ./sgd-classifier2.py

Train size: (9137, 2) | Test size: (1000, 2)
Train label distribution:
 label
ab    5115
no    1648
po     942
bo     478
le     286
se     267
ed     206
ra     105
re      90
Name: count, dtype: int64

SGD, Unigram Counts
Train score: 0.7 ; Validation score: 0.61

SGD, Unigram Tf-Idf
Train score: 0.67 ; Validation score: 0.61

SGD, Bigram Counts
Train score: 0.89 ; Validation score: 0.57

SGD, Bigram Tf-Idf
Train score: 0.86 ; Validation score: 0.6

SGD Test Result, Unigram Counts: 0.575
Error Rate: 0.42
              precision    recall  f1-score   support

          ab       0.61      0.89      0.73       550
          bo       0.64      0.11      0.19        64
          ed       0.00      0.00      0.00        21
          le       0.15      0.08      0.11        37
          no       0.42      0.17      0.24       186
          po       0.49      0.40      0.44        90
          ra       0.00      0.00      0.00        17
          re       0.50      0.27      0.35        11
 

## Updated Code for SGD Tuning

In [63]:
!cat ./sgd-tune-clasifier.py

#!/usr/bin/env python3
## SGD Classifier with Hyperparameter Tuning — fixed version
## - keep_default_na=False so labels like "NA"/"NaN" survive
## - .to_numpy() instead of .values to avoid PyArrow indexing bug
## - random_state=42 for reproducible splits and searches
## - consistent use of X_test_tfidf for the tuned bigram-tf-idf model

import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from os import system
from scipy.sparse import csr_matrix, save_npz
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.metrics import classification_report
from scipy.stats import uniform
from joblib import dump, load

# ---------------------------------------------------------------- load data
polar_train = pd.read_csv('csv/train.csv', keep_default_na=False)
polar_test  = pd.read_csv('csv/test.csv',  keep_

## Classification with SGD Tuning

In [64]:
!time python ./sgd-tune-clasifier.py

Train size: (9137, 2) | Test size: (1000, 2)
Train label distribution:
 label
ab    5115
no    1648
po     942
bo     478
le     286
se     267
ed     206
ra     105
re      90
Name: count, dtype: int64

Unigram Counts
Train score: 0.7 ; Validation score: 0.61

Unigram Tf-Idf
Train score: 0.67 ; Validation score: 0.61

Bigram Counts
Train score: 0.89 ; Validation score: 0.57

Bigram Tf-Idf
Train score: 0.86 ; Validation score: 0.6

Best params (phase 1): {'eta0': 0.0032435502100527395, 'learning_rate': 'adaptive', 'loss': 'squared_hinge'}
Best score  (phase 1): 0.6047941226844117

Best params (phase 2): {'alpha': 7.372719958564209e-05, 'penalty': 'elasticnet'}
Best score  (phase 2): 0.6036999708960774

Tuned SGD Test Result, Bigram Tf-Idf: 0.583
Error Rate: 0.42
----------
              precision    recall  f1-score   support

          ab       0.63      0.88      0.73       550
          bo       0.60      0.09      0.16        64
          ed       0.00      0.00      0.00        21

## Updated Code for Support Vector Machines

In [65]:
!cat ./svm-classifier.py

#!/usr/bin/env python3
## Support Vector Machine Classifier — fixed version
## - keep_default_na=False so labels like "NA"/"NaN" survive
## - .to_numpy() instead of .values to avoid PyArrow indexing bug
## - separate X_test_* variables for count vs tf-idf models
## - random_state=42 for reproducible splits

import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from os import system
from scipy.sparse import csr_matrix, save_npz
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import classification_report
from joblib import dump, load

# ---------------------------------------------------------------- load data
polar_train = pd.read_csv('csv/train.csv', keep_default_na=False)
polar_test  = pd.read_csv('csv/test.csv',  keep_default_na=False)

# Sanity check — abort early if labels still contain NaN
if polar_train['label'].isna()

## Classification with SVM

In [66]:
!time python ./svm-classifier.py

Train size: (9137, 2) | Test size: (1000, 2)
Train label distribution:
 label
ab    5115
no    1648
po     942
bo     478
le     286
se     267
ed     206
ra     105
re      90
Name: count, dtype: int64

SVM, Unigram Counts
Train score: 0.68 ; Validation score: 0.61

SVM, Unigram Tf-Idf
Train score: 0.63 ; Validation score: 0.6

SVM, Bigram Counts
Train score: 0.88 ; Validation score: 0.58

SVM, Bigram Tf-Idf
Train score: 0.73 ; Validation score: 0.6

SVM Test Result, Unigram Counts: 0.585
Error Rate: 0.41
              precision    recall  f1-score   support

          ab       0.60      0.91      0.72       550
          bo       1.00      0.09      0.17        64
          ed       0.00      0.00      0.00        21
          le       0.27      0.11      0.15        37
          no       0.41      0.15      0.22       186
          po       0.63      0.44      0.52        90
          ra       0.00      0.00      0.00        17
          re       0.67      0.36      0.47        11
 

## Summary

Across all five classical machine-learning baselines evaluated on the **myHateSpeech** dataset (9,137 training rows / 1,000 test rows, 9 imbalanced hate-speech classes), the overall test accuracy clustered in a narrow band between **0.51 and 0.60**, with **SGD using unigram TF-IDF features achieving the best raw accuracy (0.600)**, followed closely by **SVM (0.597)** and **Random Forest (0.592)** on the same feature representation. Decision Tree was clearly the weakest model family — its ~0.92 training score against ~0.51–0.52 test score is textbook overfitting, and even its best test configuration (0.522) trails the other learners by 6–8 accuracy points. Random Forest, SGD, and SVM all produced similar peak results within ~1 point of each other, which suggests the ceiling here is set by the data rather than the classifier: the rare hate-speech categories (`ra` = 105, `re` = 90, `ed` = 206 in training) are almost never predicted, driving macro-F1 down to ~0.25 everywhere despite weighted-F1 around 0.51–0.53. TF-IDF representations consistently edged out raw counts for the linear models (SGD, SVM), while Random Forest was more indifferent between them. The two-phase hyperparameter tuning of SGD (best: `squared_hinge` loss, `adaptive` learning rate, `eta0≈0.0032`, `elasticnet` penalty, `alpha≈7.4e-5`) **did not improve** on the untuned SGD — test accuracy actually dropped from 0.600 to 0.583 — indicating that the default `SGDClassifier` parameters were already near-optimal for this dataset and that the bottleneck is the class imbalance across the 9 hate-speech categories, not the hyperparameters. If higher accuracy is the goal, the next steps should focus on data-level techniques (class weighting, resampling, or merging rare hate-speech classes) rather than further model tuning.

## Comparison Table

| Model | Best Feature Representation | Test Accuracy | Macro F1 | Weighted F1 | Notes |
|---|---|---|---|---|---|
| Decision Tree | Unigram Counts / Bigram Counts | **0.522** | 0.24 | 0.49 | Severe overfitting (train 0.92 vs. test 0.52) |
| Random Forest | Unigram TF-IDF | **0.592** | 0.26 | 0.53 | Best non-linear model; ~16× slower than SGD |
| SGD (untuned) | Unigram TF-IDF | **0.600** | 0.25 | 0.53 | **Best overall accuracy**; fastest to train |
| SGD (tuned) | Bigram TF-IDF | 0.583 | 0.25 | 0.53 | Tuning did **not** beat untuned SGD |
| SVM (linear) | Unigram TF-IDF / Bigram TF-IDF | **0.597** | 0.25 | 0.52 | Tied best with SGD in macro performance |

### Per-Configuration Breakdown (Test Accuracy)

| Feature Representation | Decision Tree | Random Forest | SGD (untuned) | SGD (tuned) | SVM |
|---|---|---|---|---|---|
| Unigram Counts | 0.522 | 0.576 | 0.575 | — | 0.585 |
| Unigram TF-IDF | 0.515 | **0.592** | **0.600** | — | **0.597** |
| Bigram Counts | **0.522** | 0.571 | 0.553 | — | 0.574 |
| Bigram TF-IDF | 0.508 | 0.582 | 0.571 | 0.583 | **0.597** |

### Per-Configuration Breakdown (Macro F1)

| Feature Representation | Decision Tree | Random Forest | SGD (untuned) | SGD (tuned) | SVM |
|---|---|---|---|---|---|
| Unigram Counts | 0.24 | 0.24 | 0.24 | — | 0.25 |
| Unigram TF-IDF | 0.21 | 0.26 | 0.25 | — | 0.20 |
| Bigram Counts | 0.23 | 0.21 | 0.26 | — | 0.26 |
| Bigram TF-IDF | 0.22 | 0.21 | 0.27 | 0.25 | 0.22 |
